sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,191 @@
+	// for finding specific items. dataMap is kept in sync with the data slice
+	dataMap map[string]*Item
+
+	// lock is a read/write mutex, and used to facilitate read/write locks on the
+	// data and dataMap fields
+	lock sync.RWMutex
+}
+
+// queue is the internal data structure used to satisfy heap.Interface. This
+// prevents users from calling Pop and Push heap methods directly
+type queue []*Item
+
+// Item is something managed in the priority queue
+type Item struct {
+	// Key is a unique string used to identify items in the internal data map
+	Key string
+	// Value is an unspecified type that implementations can use to store
+	// information
+	Value interface{}
+
+	// Priority determines ordering in the queue, with the lowest value being the
+	// highest priority
+	Priority int64
+
+	// index is an internal value used by the heap package, and should not be
+	// modified by any consumer of the priority queue
+	index int
+}
+
+// Len returns the count of items in the Priority Queue
+func (pq *PriorityQueue) Len() int {
+	pq.lock.RLock()
+	defer pq.lock.RUnlock()
+	return pq.data.Len()
+}
+
+// Pop pops the highest priority item from the queue. This is a
+// wrapper/convenience method that calls heap.Pop, so consumers do not need to
+// invoke heap functions directly
+func (pq *PriorityQueue) Pop() (*Item, error) {
+	if pq.Len() == 0 {
+		return nil, ErrEmpty
+	}
+
+	pq.lock.Lock()
+	defer pq.lock.Unlock()
+
+	item := heap.Pop(&pq.data).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// Push pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) Push(i *Item) error {
+	if i == nil || i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.lock.Lock()
+	defer pq.lock.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
+	}
+	// copy the item value(s) so that modifications to the source item does not
+	// affect the item on the queue
+	clone, err := copystructure.Copy(i)
+	if err != nil {
+		return err
+	}
+
+	pq.dataMap[i.Key] = clone.(*Item)
+	heap.Push(&pq.data, clone)
+	return nil
+}
+
+// PopByKey searches the queue for an item with the given key and removes it
+// from the queue if found. Returns ErrItemNotFound(key) if not found. This
+// method must fix the queue after removal.
+func (pq *PriorityQueue) PopByKey(key string) (*Item, error) {
+	pq.lock.Lock()
+	defer pq.lock.Unlock()
+
+	item, ok := pq.dataMap[key]
+	if !ok {
+		return nil, nil
+	}
+
+	// remove the item the heap and delete it from the dataMap
+	itemRaw := heap.Remove(&pq.data, item.index)
+	delete(pq.dataMap, key)
+
+	if i, ok := itemRaw.(*Item); ok {
```

Comment:
might want to nil check itemRaw before type casting this.

Reply:
Good call, done in 49979c2a9